localstack/utils/cloudformation/template_deployer.py

```diff
@@ -1437,12 +1433,27 @@ def apply_parameter_changes(self, old_stack, new_stack) -> None:
 
             parameters[logical_id] = param
 
-        parameters.update({p["ParameterKey"]: p for p in new_stack.metadata["Parameters"]})
+        def _update_params(params_list: List[Dict]):
+            for param in params_list:
+                # make sure we preserve parameter values if UsePreviousValue=true
+                if not param.get("UsePreviousValue"):
```

Comment:
Definitely :+1:  on making things more immutable and avoiding re-computation. I think caching in this case is actually probably a dangerous band-aid since it just hides the underlying issues and will be fairly hard to debug. (Also not sure how invalidation would play into this)